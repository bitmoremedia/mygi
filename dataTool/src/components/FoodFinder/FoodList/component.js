import React, { PropTypes } from 'react';

const FoodList = ({ list, sourceName, sourceId, associateDataSource, deleteAssociatedDataSource }) => {

  // note: current assumption is that a food can only be associated with one item in the main Food list

  let associatedFoodItem;
  list.forEach((food)=>{
    if (food.sources && food.sources[sourceName] && food.sources[sourceName] === sourceId) {
      associatedFoodItem = food;
    }
  });

  const updateDataSource = (mode, food) => {
    if ( mode === 'add' ){
      associateDataSource({
        "foodId": food.id,
        sourceName,
        sourceId,
      });
    } else if ( mode === 'delete' ){
      deleteAssociatedDataSource({
        "foodId": food.id,
        sourceName,
        sourceId,
      });
    }
  }

  const renderAction = (food) => {
    if (associatedFoodItem) {
      if (food.id === associatedFoodItem.id) {
        return <button onClick={()=>updateDataSource('delete', food)}> - </button>;
      }
      return '';
    }
    return <button onClick={()=>updateDataSource('add', food)}> + </button>;
  };

  return (
    <div>
      <ul>
        {list.map((food)=><li key={food.id}>{food.name} [{food.gi}] {renderAction(food)}</li>)}
      </ul>
    </div>
  );
};

FoodList.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      gi: PropTypes.any.isRequired,
    })
  ).isRequired,
};

export default FoodList;
