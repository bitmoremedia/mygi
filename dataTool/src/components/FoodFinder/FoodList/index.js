import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { associateDataSource, deleteAssociatedDataSource } from '../../../redux/actions'

import FoodList from './component'

const mapStateToProps = () => {}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ associateDataSource, deleteAssociatedDataSource }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(FoodList)
