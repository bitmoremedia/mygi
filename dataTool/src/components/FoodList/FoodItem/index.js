import React, { PropTypes } from 'react';

import { Item } from './styles';

const FoodItem = ({food, editAction}) => {
  return (
    <Item key={`${food.id}-food-item`}>
      {food.name}
    </Item>
  );
};

FoodItem.propTypes = {
  food: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    gi: PropTypes.any.isRequired,
  }).isRequired,
};

export default FoodItem;
