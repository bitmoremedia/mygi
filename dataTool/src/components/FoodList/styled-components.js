import styled from 'styled-components'

export const AddButton = styled.div`
  margin-top: 5px;
  margin-bottom: 10px;
`
