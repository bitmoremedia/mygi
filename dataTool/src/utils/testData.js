const testData = {
  foodList: {
    'abcd-123': {
      category: 'Beans',
      gi: 35,
      id: 'abcd-123',
      name: 'Kidney Beans',
      sources: {},
    },
    'efgh-456': {
      category: 'Bread',
      gi: 68,
      id: 'efgh-456',
      name: 'White Bread',
      sources: {
        havard: 'havard-white-bread',
      },
    },
  },
  dataSources: {
    havard: {
      name: 'havard',
      title: 'Havard',
      data: {
        'havard-white-bread': {
          gi: 65,
          id: 'havard-white-bread',
          name: 'White Bread',
        },
      },
    },
    mendosa: {
      name: 'mendosa',
      title: 'Mendosa',
      data: {
        'mendosa-white-bread': {
          gi: 77,
          id: 'mendosa-white-bread',
          name: 'White Bread',
        },
      },
    },
  },
  foodId: 'mendosa-white-bread',
  sourceName: 'mendosa',
}

export default testData
