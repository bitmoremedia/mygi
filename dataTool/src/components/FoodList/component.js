import React, { Component } from 'react';
import _values from 'lodash/values';
import _sortBy from 'lodash/sortBy';
import { Link } from 'react-router-dom';

import Table from '../common/Table';
import Modal from '../common/Modal';
import AddEditFood from '../AddEditFood';
import Button from '../common/Button';

import { AddButton } from './styles';
import FoodItem from './FoodItem';

class FoodList extends Component {

  constructor(props, context) {
    super(props, context);
    this.state = {
      showAddFood: false,
      showEditFoodModal: false,
      editFoodId: undefined,
    };
    this.toggleAddFood = this.toggleAddFood.bind(this);
    this.openEditFoodModal = this.openEditFoodModal.bind(this);
    this.closeEditFoodModal = this.closeEditFoodModal.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    // a bit hacky - but if props have changed then hide the add/edit fields
    this.setState({
      showAddFood: false,
      showEditFoodModal: false,
    });
  }

  toggleAddFood() {
    this.setState({
      showAddFood: !this.state.showAddFood
    });
  }

  openEditFoodModal(id) {
    this.setState({
      showEditFoodModal: true,
      editFoodId: id
    });
  }

  closeEditFoodModal() {
    this.setState({
      showEditFoodModal: false,
      editFoodId: undefined
    });
  }

  render() {
    const { toggleAddFood, openEditFoodModal, closeEditFoodModal } = this;
    const { showAddFood, showEditFoodModal, editFoodId } = this.state;
    const { foodList, dataSources } = this.props;
    const foodListArray = _sortBy(_values(foodList), 'name');
    const dataSourcesArray = _sortBy(_values(dataSources), 'title');

    const columns = [{
      key: 'food',
      label: 'Food'
    },{
      key: 'category',
      label: 'Category'
    },{
      key: 'gi',
      label: 'GI Value'
    },{
      key: 'average',
      label: 'Average GI'
    }];
    // add columns for each of our data sources
    dataSourcesArray.forEach((source)=>{
      columns.push({
        key: source.name,
        label: <Link to={`/source/${source.name}`}>{source.title}</Link>,
      });
    });
    const data = foodListArray.map((food)=>{
      const dataColumns = [
        { key: food.id, value: <FoodItem food={food} editAction={openEditFoodModal} /> },
        { key: food.id, value: food.category },
        { key: food.id, value: food.gi },
        { key: food.id, value: 'TBC' },
      ];
      let giTotal = 0, giCount = 0, giAverage;
      // add data columns for each of our data sources
      dataSourcesArray.forEach((source)=>{
        const sourceGi = (food.sources[source.name] && source.data[food.sources[source.name]]) ? source.data[food.sources[source.name]].gi : "";
        dataColumns.push({
          key: `${food.id}-${source.name}`,
          value: sourceGi,
        });
        if (parseInt(sourceGi, 10) > 0) {
          giTotal += parseInt(sourceGi, 10);
          giCount += 1;
        }
      });
      // update the average GI value if we have one
      if ( giTotal > 0 && giCount > 0 ){
         giAverage = (giTotal/giCount);
        dataColumns[3].value = giAverage.toFixed(0);
      }
      return dataColumns;
    });

    const renderToggleAddFoodButton = () => {
      const buttonText = (this.state.showAddFood) ? 'Cancel' : 'Add';
      return <AddButton><Button onClick={()=>toggleAddFood()}>{buttonText}</Button></AddButton>;
    }

    return (
      <div>
        {renderToggleAddFoodButton()}
        {showAddFood && <AddEditFood layout="inline"/>}
        { foodListArray.length > 0 && dataSourcesArray.length > 0 && <Table columns={columns} data={data}/> }
        <Modal visible={showEditFoodModal} onClose={closeEditFoodModal}>
          {showEditFoodModal && <AddEditFood mode='edit' editFoodId={editFoodId} />}
        </Modal>
      </div>
    );
  }
}

export default FoodList;
