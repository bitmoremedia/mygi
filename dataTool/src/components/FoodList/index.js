import React, { Component } from 'react';

import { getFoodList, getDataSources, getCategories } from '../../api';
import Table from '../common/Table';
import FoodItem from '../FoodItem';

class FoodList extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      foodList: [],
      dataSources: [],
      categories: [],
    };
  }

  componentDidMount(){
    getFoodList()
      .then((foodList)=>{this.setState({foodList});})
      .catch((err)=>{console.log(err);});

    getDataSources()
      .then((dataSources)=>{this.setState({dataSources});})
      .catch((err)=>{console.log(err);});

    getCategories()
      .then((categories)=>{this.setState({categories});})
      .catch((err)=>{console.log(err);});
  }

  renderFoodItems(item) {
    const { foodList, dataSources } = this.state;
    const columns = [{
      key: 'food',
      label: 'Food'
    },{
      key: 'gi',
      label: 'GI Value'
    }];
    // add columns for each of our data sources
    dataSources.forEach((source)=>{
      columns.push({
        key: source.name,
        label: source.title,
      });
    });

    const data = foodList.map((food)=>{
      const dataColumns = [
        { key: food.id, value: <FoodItem food={food}/> },
        { key: food.id, value: food.gi }
      ];
      // add data columns for each of our data sources
      dataSources.forEach((source)=>{
        const sourceGi = (food.sources[source.name]) ? food.sources[source.name].gi : "";
        dataColumns.push({
          key: `${food.id}-${source.name}`,
          value: sourceGi,
        });
      });
      return dataColumns;
    });
    return (
      <Table columns={columns} data={data}/>
    );
  }

  render() {
    return (
      <div>
        {this.renderFoodItems()}
      </div>
    );
  }
}

export default FoodList;
