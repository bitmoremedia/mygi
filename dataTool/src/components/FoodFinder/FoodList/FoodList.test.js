import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import _values from 'lodash/values'

import testData from '../../../utils/testData'

import FoodList from './component'

const foodListArray = _values(testData.foodList)

test('component renders without error', () => {
  const wrapper = mount(
    <FoodList
      list={foodListArray}
      sourceName={testData.sourceName}
      sourceId={testData.foodId}
      associateDataSource={() => {}}
      deleteAssociatedDataSource={() => {}}
    />,
  )
  expect(wrapper.exists()).toBe(true)
})

// SNAPSHOT TESTS

test('component renders as expected', () => {
  const wrapper = mount(
    <FoodList
      list={foodListArray}
      sourceName={testData.sourceName}
      sourceId={testData.foodId}
      associateDataSource={() => {}}
      deleteAssociatedDataSource={() => {}}
    />,
  )
  expect(toJson(wrapper)).toMatchSnapshot()
})
