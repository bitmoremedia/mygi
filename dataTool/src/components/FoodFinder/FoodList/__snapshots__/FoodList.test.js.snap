exports[`test component renders as expected 1`] = `
<Component
  associateDataSource={[Function]}
  deleteAssociatedDataSource={[Function]}
  list={
    Array [
      Object {
        "category": "Beans",
        "gi": 35,
        "id": "abcd-123",
        "name": "Kidney Beans",
      },
      Object {
        "category": "Bread",
        "gi": 68,
        "id": "efgh-456",
        "name": "White Bread",
        "sources": Object {
          "havard": "havard-white-bread",
        },
      },
    ]
  }
  sourceId="mendosa-white-bread"
  sourceName="mendosa">
  <div>
    <ul>
      <li>
        Kidney Beans
         [
        35
        ] 
        <Component
          onClick={[Function]}
          type="add">
          <styled.span
            onClick={[Function]}>
            <span
              className="bBeLnM"
              onClick={[Function]}>
              Assign
            </span>
          </styled.span>
        </Component>
      </li>
      <li>
        White Bread
         [
        68
        ] 
        <Component
          onClick={[Function]}
          type="add">
          <styled.span
            onClick={[Function]}>
            <span
              className="bBeLnM"
              onClick={[Function]}>
              Assign
            </span>
          </styled.span>
        </Component>
      </li>
    </ul>
  </div>
</Component>
`;
